core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMove.java

```diff
@@ -0,0 +1,208 @@
+package org.optaplanner.core.impl.heuristic.selector.move.generic.list.kopt;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.api.function.TriPredicate;
+import org.optaplanner.core.api.score.director.ScoreDirector;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.index.IndexVariableSupply;
+import org.optaplanner.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
+import org.optaplanner.core.impl.heuristic.move.AbstractMove;
+import org.optaplanner.core.impl.heuristic.move.Move;
+import org.optaplanner.core.impl.score.director.InnerScoreDirector;
+
+/**
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ * @param <Node_> The value type
+ */
+public class KOptListMove<Solution_, Node_> extends AbstractMove<Solution_> {
+
+    private final ListVariableDescriptor<Solution_> listVariableDescriptor;
+    private final Object entity;
+    private final KOptDescriptor<Solution_, Node_> descriptor;
+    private final List<FlipSublistMove<Solution_>> equivalent2Opts;
+    private final int postShiftAmount;
+
+    public KOptListMove(ListVariableDescriptor<Solution_> listVariableDescriptor,
+            Object entity,
+            KOptDescriptor<Solution_, Node_> descriptor,
+            List<FlipSublistMove<Solution_>> equivalent2Opts,
+            int postShiftAmount) {
+        this.listVariableDescriptor = listVariableDescriptor;
+        this.entity = entity;
+        this.descriptor = descriptor;
+        this.equivalent2Opts = equivalent2Opts;
+        this.postShiftAmount = postShiftAmount;
+    }
+
+    @Override
+    protected AbstractMove<Solution_> createUndoMove(ScoreDirector<Solution_> scoreDirector) {
+        if (equivalent2Opts.isEmpty()) {
+            return this;
+        } else {
+            List<FlipSublistMove<Solution_>> inverse2Opts = new ArrayList<>(equivalent2Opts.size());
+            for (int i = equivalent2Opts.size() - 1; i >= 0; i--) {
+                inverse2Opts.add(equivalent2Opts.get(i).createUndoMove(scoreDirector));
+            }
+            return new UndoKOptListMove<>(listVariableDescriptor, entity, descriptor, inverse2Opts, -postShiftAmount);
+        }
+    }
+
+    @Override
+    protected void doMoveOnGenuineVariables(ScoreDirector<Solution_> scoreDirector) {
+        equivalent2Opts.forEach(move -> move.doMoveOnGenuineVariables(scoreDirector));
```

Comment:
This is really a nitpick now, but due to `scoreDirector`, this is now a capturing lambda; therefore a new instance is created every time the method is run on the hot path.
I suggest replacing with a normal loop, same in the undo move.